pkg/segment/query/processor/dataprocessor_test.go

```diff
@@ -256,3 +277,251 @@ func Test_Fetch_multipleStreams(t *testing.T) {
 	_, err := dp.Fetch()
 	assert.Equal(t, io.EOF, err)
 }
+
+func Test_Fetch_multipleBottleneck(t *testing.T) {
+	stream := &mockStreamer{
+		allRecords: map[string][]utils.CValueEnclosure{
+			"col1": {
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "a"},
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "b"},
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "c"},
+			},
+		},
+		qid: 0,
+	}
+
+	dp0 := &DataProcessor{
+		streams:   []*cachedStream{{stream, nil, false}},
+		processor: &passThroughProcessor{},
+	}
+
+	dp1 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp0)},
+		processor:       &mockBottleneckProcessor{name: "dp1"},
+		isBottleneckCmd: true,
+		isTwoPassCmd:    false,
+	}
+
+	dp2 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp1)},
+		processor:       &mockBottleneckProcessor{name: "dp2"},
+		isBottleneckCmd: true,
+		isTwoPassCmd:    false,
+	}
+
+	output, err := dp2.Fetch()
+	assert.Equal(t, io.EOF, err)
+	assert.NotNil(t, output)
+	assert.Equal(t, 1, dp2.processor.(*mockBottleneckProcessor).numSeen)
+}
+
+func Test_Fetch_multipleBottleneck_inputNil(t *testing.T) {
+	stream := &mockStreamer{
+		allRecords: map[string][]utils.CValueEnclosure{},
+		qid:        0,
+	}
+
+	dp1 := &DataProcessor{
+		streams:   []*cachedStream{{stream, nil, false}},
+		processor: &passThroughProcessor{},
+	}
+
+	dp2 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp1)},
+		processor:       &mockBottleneckProcessor{name: "dp2"},
+		isBottleneckCmd: true,
+		isTwoPassCmd:    false,
+	}
+
+	dp3 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp2)},
+		processor:       &mockBottleneckProcessor{name: "dp3"},
+		isBottleneckCmd: true,
+		isTwoPassCmd:    false,
+	}
+
+	output, err := dp3.Fetch()
+	assert.Equal(t, io.EOF, err)
+	assert.Nil(t, output)
+	assert.Equal(t, 0, dp3.processor.(*mockBottleneckProcessor).numSeen)
+}
+
+func Test_Fetch_multipleBottleneck_twoPass(t *testing.T) {
+	stream := &mockStreamer{
+		allRecords: map[string][]utils.CValueEnclosure{
+			"col1": {
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "a"},
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "b"},
+				utils.CValueEnclosure{Dtype: utils.SS_DT_STRING, CVal: "c"},
+			},
+		},
+	}
+
+	dp0 := &DataProcessor{
+		streams:   []*cachedStream{{stream, nil, false}},
+		processor: &passThroughProcessor{},
+	}
+
+	dp1 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp0)},
+		processor:       &mockBottleneckProcessor{name: "dp1"},
+		isBottleneckCmd: true,
+	}
+
+	dp2 := &DataProcessor{
+		streams:         []*cachedStream{NewCachedStream(dp1)},
+		processor:       &mockBottleneckProcessor{name: "dp2"},
+		isBottleneckCmd: true,
+	}
+
+	dp3 := &DataProcessor{
+		streams:      []*cachedStream{NewCachedStream(dp2)},
+		processor:    &mockTwoPassProcessor{name: "dp3"},
+		isTwoPassCmd: true,
+	}
```

Comment:
You should probably also have `isBottleneckCmd: true`, since a two-pass command is always a bottleneck